BIP-112: Mempool-only CHECKSEQUENCEVERIFY

maaku opened the pull request:
Replace NOP3 with CHECKSEQUENCEVERIFY (BIP-112)

&lt;nSequence&gt; CHECKSEQUENCEVERIFY -> &lt;nSequence&gt;

Fails if relative lock-time encoded in txin.nSequence < relative lock-time represented by nSequence, allowing funds of a txout to be locked for a number of blocks or a duration of time after its inclusion in a block.

Transactions that fail CSV verification will be rejected from the mempool, making it easy to test the feature. However blocks containing "invalid" CSV-using transactions will still be accepted; this is _not_ the soft-fork required to actually enable CSV for production use.

maaku committed:
Replace NOP3 with CHECKSEQUENCEVERIFY (BIP112)

<nSequence> CHECKSEQUENCEVERIFY -> <nSequence>

Fails if txin.nSequence < nSequence, allowing funds of a txout to be locked for a number of blocks or a duration of time after its inclusion in a block.

Only the logic and unit tests are implemented; this commit does not have changes to policy or consensus behavior.

maaku committed:
Add CTxIn::SEQUENCE_LOCKTIME_DISABLED_FLAG, CTxIn::SEQUENCE_LOCKTIME_SECONDS_FLAG, and CTxIn::SEQUENCE_LOCKTIME_MASK to primitive/transaction.h

Will be needed by CHECKSEQUENCEVERIFY code.

maaku committed:
Enable CHECKSEQUENCEVERIFY as a standard script verify flag

Transactions that fail CSV verification will be rejected from the mempool, making it easy to test the feature. However blocks containing "invalid" CSV-using transactions will still be accepted; this is *not* the soft-fork required to actually enable CSV for production use.

btcdrak commented:
The BIP text is available at https://github.com/bitcoin/bips/pull/179

kanzure commented:
Should mempools always reject CSV-failing transactions?

dcousens commented on src/script/interpreter.cpp:
```diff
@@ -1126,8 +1166,30 @@ bool TransactionSignatureChecker::CheckSig(const vector<unsigned char>& vchSigIn
+        // relative lock-time.
+        txToLockTime = (int64_t)~txTo->vin[nIn].nSequence;
+        // Sequence numbers under SEQUENCE_THRESHOLD are not consensus
+        // constrained.
```
Could this please read:  `Inverted sequence numbers above (or equal to) SEQUENCE_THRESHOLD are not consensus constrained`

dcousens commented on src/script/interpreter.cpp:
```diff
@@ -1126,8 +1166,30 @@ bool TransactionSignatureChecker::CheckSig(const vector<unsigned char>& vchSigIn
+    // in lock time to the sequence number of the input. All other
+    // logic is the same, all that differs is what we are comparing
+    // the lock time to.
+    int64_t txToLockTime;
```
Since this is repeatedly referred to as a relative lock time,  can we reflect that? `relativeLockTime`?

dcousens commented on src/script/interpreter.cpp:
```diff
@@ -1126,8 +1166,30 @@ bool TransactionSignatureChecker::CheckSig(const vector<unsigned char>& vchSigIn
 }
 
-bool TransactionSignatureChecker::CheckLockTime(const CScriptNum& nLockTime) const
+bool TransactionSignatureChecker::CheckLockTime(const CScriptNum& nLockTime, bool fSequence) const
```
Why is this being put in the same function anyway?
It only shares 10 lines of code,  and adds another 10.
It could be easily a 10 line function and a 16 line function without an `fSequence` parameter, and no need to have two sets of comments dictating two different use cases.

maaku commented:
@kanzure This pull request is implemented such that CSV-failing transactions with tx.nVersion>=2 are considered non-standard and therefore rejected from mempool and/or relay. Given that CSV re-purposes a presently unused NOP opcode, there shouldn't be any compatibility issues. Is there a case you have in mind where a CSV-failing transaction should be added to the mempool?

maaku committed:
Pull most of CheckLockTime() out into VerifyLockTime(), a local function that will be reused for CheckSequence()

maaku commented:
Rebased, and made compliant with latest version of #6312. Also merged @dcousens code duplication nit, and split the main commit into two for easier review.

Ready for review and merge.

btcdrak commented on src/script/interpreter.cpp:
```diff
@@ -379,7 +379,45 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
+                    if (nSequence >= CTxIn::SEQUENCE_LOCKTIME_THRESHOLD)
+                        break;
+
+                    // Actually compare the specified inverse sequence number
```
s/inverse//

maaku replied:
This email I got! Thanks I make the correction.
On Oct 1, 2015 11:12 AM, "฿tcDrak" <notifications@github.com> wrote:

> In src/script/interpreter.cpp
> <https://github.com/bitcoin/bitcoin/pull/6564#discussion_r40946716>:
>
> > +                    // 5-byte numeric operands.
> > +                    const CScriptNum nSequence(stacktop(-1), fRequireMinimal, 5);
> > +
> > +                    // In the rare event that the argument may be < 0 due to
> > +                    // some arithmetic being done first, you can always use
> > +                    // 0 MAX CHECKSEQUENCEVERIFY.
> > +                    if (nSequence < 0)
> > +                        return set_error(serror, SCRIPT_ERR_NEGATIVE_LOCKTIME);
> > +
> > +                    // To provide for future soft-fork extensibility, if the
> > +                    // operand is too large to be treated as a relative lock-
> > +                    // time, CHECKSEQUENCEVERIFY behaves as a NOP.
> > +                    if (nSequence >= CTxIn::SEQUENCE_LOCKTIME_THRESHOLD)
> > +                        break;
> > +
> > +                    // Actually compare the specified inverse sequence number
>
> s/inverse//
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6564/files#r40946716>.
>

kanzure commented on src/test/data/tx_valid.json:
```diff
@@ -229,5 +229,87 @@
+[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "4294967296 NOP3 1"]],
+"020000000100010000000000000000000000000000000000000000000000000000000000000000000000ffffffff0100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
+
+["5 byte non-minimally-encoded operandss are valid"],
```
nit: typo ("operandss")

kanzure commented:
utACK

btcdrak commented:
needs rebase

petertodd commented:
These changes would be easier to review if they were built on top of  #6312 I think, showing the final state of the new CSV feature.

petertodd commented:
Or, am I meant to be reviewing #6566? Rather confusing.

rustyrussell commented:
utAck

This doesn't seem to make sense without BIP68.  I tried to test it, got confused as it errored out comparing a BIP68 nSequence (1073742784, ie 0x400003C0) with a locktime of 500000030....

petertodd commented:
Again, I think these pull-req's need to be reviewed as a single unit;
they don't make sense apart from each other.

btcdrak commented:
needs rebase

btcdrak commented:
I'm leaning to agree with @petertodd about merging #6312 into this PR (or vice versa). #6312 has become quite polished, there are a couple of minor nits to polish off. It seems like the right time to merge both together now since they make sense as a package anyway.

jtimon commented:
I agree it's hard to review #6566, #6564 and #6312 like this. What is supposed to come first?
If you want to maintain t\several PRs that's fine. But at least you could rebase one on top of the other.
What advantage there is in having them be completely "independent" from each other?
We will have a preferred merge order, right?

btcdrak commented:
@jtimon #6312 (BIP68) and this PR are supposed to be independent of each other, but I think that's been pedantic since OP_CSV isn't useful without BIP68. Merge order doesnt matter. 

I think it made sense to have two PRs while narrowing down the semantics of BIP68, but now that's done, it's better to have 1 PR. It will be more expedient for review if we rebase BIP68 onto this PR.

For reference, #6566 is independent and at the last meeting we discussed aiming to merge #6566 for inclusion with the BIP65 soft fork and for BIP68/BIP112 to be included in a separate release.

jtimon commented:
> Merge order doesn't matter.

Mhmm, that's strange to me, but anyway...Then let's just pick one and rebase one of top of the other. Do we gain anything from not deciding the merge order in advance and maintaining the 3 PRs independent?

maaku commented:
Consistent feedback from the bitcoin pull request process has been to break up pulls to do one minimal thing, to not have multi-feature pulls, and to not have dependent pull requests when at all possible. In light of this and since this pull and #6312 don't actually share any code (except for 3 constants in `src/primitives/transaction.h`), they were submitted separately.

btcdrak commented:
Needs rebase.

maaku committed:
Add bitwise AND operator to CScriptNum

maaku commented:
Rebased with new semantics.

btcdrak commented:
utACK

instagibbs commented on src/script/script.h:
```diff
@@ -251,6 +252,11 @@ class CScriptNum
+    inline CScriptNum operator&(   const int64_t& rhs)    const { return CScriptNum(m_value & rhs);}
+    inline CScriptNum operator&(   const CScriptNum& rhs) const { return operator&(rhs.m_value);   }
+
+    inline CScriptNum& operator&=( const CScriptNum& rhs)       { return operator+=(rhs.m_value);  }
```
Should be &= not += I assume

instagibbs commented:
utACK otherwise

dcousens commented:
utACK

rubensayshi commented:
utACK

jmcorgan commented on src/primitives/transaction.h:
```diff
@@ -61,6 +61,18 @@ class CTxIn
 
+    /* If this flag set, CTxIn::nSequence is NOT interpreted as a
+     * relative lock-time. Setting the most significant bit of a
+     * sequence number disabled relative lock-time. */
```
s/disabled/disables/

afk11 commented:
utACK, will bump this to tACK later today..

jmcorgan commented:
utACK

dcousens commented:
Needs rebase

btcdrak committed:
Typo fixes and added comments

btcdrak commented:
A couple of people have requested to see the #6312 and #6564 rebased together in one branch for review. I have done this at https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv. Please note there is one less commit 78370bb827320660e2016ea7e42fedc5a66677dd was duplicated in both. I'm happy to roll that into a PR if necessary, although the purpose is really just to assist the testers who asked for it.

btcdrak commented:
The BIP68+OP_CSV combined branch has been updated https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv

jgarzik commented:
ut ACK

greenaddress commented:
utACK

btcdrak committed:
Rename flags

_SECONDS_FLAG to _TYPE_FLAG
_DISABLED_FLAG to _DISABLE_FLAG

btcdrak commented:
BIP text updated in https://github.com/bitcoin/bips/pull/248

btcdrak committed:
Fixup comments

afk11 commented on src/script/interpreter.cpp:
```diff
@@ -370,7 +370,45 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
+                    if ((nSequence & CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0)
+                        break;
+
+                    // Actually compare the specified inverse sequence number
```
We're not inverting the sequence number any more, should this line clarified/removed?

btcdrak replied:
Fixed.

rustyrussell commented:
Tested-by: Rusty Russell <rusty@rustcorp.com.au>

(I merged both this and #6312  for the test, with only a trivial conflict to resolve).

ajtowns commented:
tested ACK (combined with #7184, #7187). https://github.com/ajtowns/op_csv-test -- lightning-style HTLCs using OP_CSV seem to perform as expected

morcos commented:
Personally, I would find this code cleaner and safer if VerifyLockTime was not encapsulated for reuse for nLockTime and nSequence checks.  The amount of repeated code is minimal and it would save the fact that we have to jump through a couple of hoops to reuse it now.

That said, if I'm the only one who feels this way, don't let me stand in the way.

petertodd commented:
@morcos I'll second being dubious about reuse.

Anyway, ecosystem wide all this stuff will get re-written easily a dozen times in the various libraries...

maaku commented:
I think you both missed the point I was trying to make. I'm not saying the
code should be unified, just pointing that out as an example of something
that would be made more difficult by differing semantics. For better or
worse the semantics of nLockTime are already firmly established, and it is
better that both locks use the same semantics as that is less likely to
cause problems than otherwise. Even if the existing semantics are a bit
weird and not what I would have done, at least they'll be consistent.
On Feb 2, 2016 9:28 AM, "Peter Todd" <notifications@github.com> wrote:

> @morcos <https://github.com/morcos> I'll second being dubious about reuse.
>
> Anyway, ecosystem wide all this stuff will get re-written easily a dozen
> times in the various libraries...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-178701138>.
>

morcos commented:
@maaku I wasn't necessarily conflating these two things.  I'm ok leaving the semantics for BIP 68 the same as they are now (and same as nLockTime).  I can't decide myself which option I like better.  

Separately I think, even if we leave the semantics the same in #7184, that this PR would be better served by not reusing VerifyLockTime in both places.

btcdrak commented:
In any case, this is just a differing opinion about implementation, which is a non-blocking nit.

josephpoon commented:
Tested ACK using btcdrak's BIP68+OP_CSV combined branch https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv on regtest.

This pull request is _very_ useful for Lightning Network channels without pre-set expiries. Thanks~~~!

maaku commented:
This PR will be updated as soon as BIP 68 is merged.
On Feb 2, 2016 8:25 PM, "josephpoon" <notifications@github.com> wrote:

> Tested ACK using btcdrak's BIP68+OP_CSV combined branch
> master...btcdrak:sequenceandcsv
> <https://github.com/bitcoin/bitcoin/compare/master...btcdrak:sequenceandcsv>
> on regtest.
>
> This pull request is *very* useful for Lightning Network channels without
> pre-set expiries. Thanks~~~!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6564#issuecomment-178998124>.
>

CodeShark commented:
Tested ACK using btcdrak's BIP68+OP_CSV combined branch master...btcdrak:sequenceandcsv on regtest.

Roasbeef commented:
[Tested ACK](https://github.com/bitcoin/bitcoin/pull/7184#issuecomment-180681158).

NicolasDorier commented:
Tested ACK and as well replicated in NBitcoin (tests included)

maaku commented:
Closing in favor of a future PR based on #7184 .

NicolasDorier commented:
Indeed commented on the wrong PR...

btcdrak commented:
@maaku Why would you close a PR that has been getting heavy actual _tested_ review? *sigh*

jtimon commented:
@btcdrak I don't see any lost in closing PR.
PRs are just like branches: petnames (in the case, in decimal) for commits.
A link to the new PR and traceabilty problems are solved.
Tomorrow I'll reset HEAD^^^^^^^^ and commit a new PR with my name on it in a single commit. Once everybody has reviewed it there's no good reason why anybody would want to know who wrote the original commits or when. Trace-what?

petertodd commented:
@jtimon "there's no good reason why anybody would want to know who wrote the original commits or when" <- careful there, we should make sure all authors of the code in question are credited appropriately!

While I haven't run into this too often yet, I'd even make a point of crediting people if I build upon their pull-reqs with total rewrites of the code; even in cases where 100% of the original code gets thrown out such people often - if not always - have done valuable work exploring the problem and deserve credit for it.

laanwj commented:
Of course people should be properly credited for their work! Why is this even a discussion?
Mentioning someone's name in the commit message and pull request is a good way to do this.

btcdrak commented:
I think i'll just reopen this in a separate PR and push the original work rebased.

maaku commented:
It was closed because there is no way I know of to change the repo a PR draws from, I will not be maintaining this patch, and for security reasons I can no longer be adding outside collaborators on my own repo to maintain these pulls. Someone who will be maintaining this patch needs to open a new PR. Blame Github.

btcdrak commented:
@maaku sure, I'm just rebasing it and will open a new one shortly.
